geomstats/stratified_geometry/wald_space.py

```diff
@@ -0,0 +1,1024 @@
+        sort_key = sorted(range(len(seq)), key=seq.__getitem__)
+        self.partition = tuple([partition[key] for key in sort_key])
+        self.split_sets = tuple([tuple(sorted(split_sets[key])) for key in sort_key])
+
+        self.where = {s: i for i, s in enumerate(self.flatten(self.split_sets))}
+
+        lengths = [len(splits) for splits in self.split_sets]
+        self.sep = [0] + [sum(lengths[0:j]) for j in range(1, len(lengths) + 1)]
+
+        self._paths = None
+        self._support = None
+        self._chart_gradient = None
+
+    @staticmethod
+    def flatten(x):
+        """Flatten a list of lists into a single list by concatenation.
+
+        Parameters
+        ----------
+        x : nested list
+            The nested list to flatten.
+
+        Returns
+        -------
+        x_flat : list, tuple
+            The flatted list.
+        """
+        return [y for z in x for y in z]
+
+    def unflatten(self, x):
+        """Transform list into list of lists according to separators, ``self.sep``.
+
+        The separators are a list of integers, increasing. Then, all elements between to
+        indices in separators will be put into a list, and together, all lists give a
+        nested list.
+
+        Parameters
+        ----------
+        x : iterable
+            The flat list that will be nested.
+
+        Returns
+        -------
+        x_nested : list[list]
+            The nested list of lists.
+        """
+        return [x[i:j] for i, j in zip(self.sep[:-1], self.sep[1:])]
+
+    def paths(self):
+        """For each pair of labels, give the unique path of splits between those labels.
+
+        A list of dictionaries, each dictionary is for the respective connected
+        component of the forest, and the items of each dictionary are for each pair
+        of labels u, v, u < v in the respective component, a list of the splits on the
+        unique path between the labels u and v.
+
+        Returns
+        -------
+        label_paths : list of dict
+            The list of dictionaries containing list of splits on path between labels.
+        """
+        if self._paths is not None:
+            return self._paths
+        self._paths = [
+            {
+                (u, v): [s for s in splits if s.separates(u, v)]
+                for u, v in it.combinations(part, r=2)
+            }
+            for part, splits in zip(self.partition, self.split_sets)
+        ]
+        return self._paths
+
+    def corr(self, x):
+        """Compute the correlation matrix of the topology with edge weights ``x``.
+
+        Parameters
+        ----------
+        x : array-like
+            Takes a vector of length 'number of total splits' of the structure.
+
+        Returns
+        -------
+        corr : array-like, shape=[n, n]
+            Returns the corresponding correlation matrix.
+        """
+        corr = np.zeros((self.n, self.n))
+        for path_dict in self.paths():
```

Comment:
What do you think about making `paths` a property? I'm not sure about this suggestion, but `self.paths` seems more natural than `self.paths()`.